design-proposals/klip-43-timestamp-data-type-support.md

```diff
@@ -0,0 +1,194 @@
+
++------------------------+
+|time                    |
++------------------------+
+|1994-11-05T13:15:30Z    |
+```
+
+TIMESTAMPS can be represented by milliseconds from Unix epoch or by using conversion functions:
+
+```roomsql
+INSERT INTO stream_name VALUES (1605927509166);
+INSERT INTO stream_name VALUES (STRINGTOTIMESTAMP("2020-11-20", "YYYY-MM-DD"));
+```
+
+## Design
+
+### Serialization/Deserialization
+
+TIMESTAMPs will be handled by the `java.time.Instant` class within KSQL. The corresponding Kafka Connect type is
+[org.apache.kafka.connect.data.Timestamp](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Timestamp.html).
+They are represented as long types in Schema Registry, but also come with a tag indicating that it
+is a timestamp, so they should be distinguishable from long types when handling serialized values in KSQL.
+
+#### Avro
+
+Avro schemas represent timestamps as
+```json
+{
+  "type": "long",
+  "logicalType": "timestamp-millis"
+}
+```
+The Avro deserializer KSQL uses supports this.
+
+#### Protobuf
+Protobuf 3 has a Timestamp type. The Protouf deserializer KSQL uses supports this.
+
+#### JSON/Delimited
+
+Timestamps will get stored in JSON and CSV files as long values. The KSQL JSON and delimited deserializers
+will be updated to parse timestamps.
+
+### UDFs
+
+The following UDFs should be updated to use the TIMESTAMP type instead of BIGINT:
+
+* TIMESTAMPTOSTRING
+* STRINGTOTIMESTAMP
+
+Because BIGINTs will be implicitly cast into TIMESTAMPs and vice versa, queries using these functions
+with BIGINT will still work when TIMESTAMP is introduced.
+
+There will also be a new function, `NOW()` that returns the current timestamp.
+
+There are a few existing UDFs that deal with dates. These should be left as is until a DATE type is implemented:
+
+* UNIX_DATE
+* DATETOSTRING
+* STRINGTODATE
+
+### Casting
+
+Casting from TIMESTAMP to BIGINT will return the millisecond representation of the timestamp, and
+casting from BIGINT to TIMESTAMP will return a TIMESTAMP that is the BIGINT number of milliseconds
+from Unix epoch. 
+
+If a user attempts to cast a negative number or a timestamp before Unix epoch, the CAST will throw
+an error.
+
+### Arithmetic operations and comparisons
+
+At the very least, we would want KSQL to be able to compare, add and subtract TIMESTAMPS.
+
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and [MariaDB](https://mariadb.com/kb/en/date-time-functions/)
+have very complete lists of built-in arithmetic functions, while [PostgreSQL](https://www.postgresql.org/docs/9.0/functions-datetime.html)
+defines behaviors for arithmetic operators.
+
+Because functions are more well-defined, we should opt for built-in arithmetic functions
+instead of using operators. The following functions are necessary:
+
+* `TIMESTAMPADD(time_stamp, big_int)` - adds `big_int` milliseconds to `time_stamp` and returns the result as a
```

Comment:
Mysql and Snowflake uses the `TIMESTAMPADD(unit, interval, datetime_expr)` format. I don't know about other DBs yet, but should we align with those systems? If so, that means we cannot handle this `timestampadd` and `timestampsub` in this klip because time units are out of scope, right?

Reply:
Yeah those syntaxes make a lot more sense. I do think that supporting timestamp arithmetic is valuable, so I might include this and time units in scope as another milestone for a later release.